### PR TITLE
feat: Add support for Category entity and implement missing functionality

### PR DESCRIPTION
**Pull request by VKhmil**
This commit adds support for the Category entity and implements the missing functionality according to the provided requirements.

User Use Cases:

Implemented category browsing functionality for users by providing endpoints to retrieve all categories and books by a specific category. Admin Use Cases:

Implemented category management functionality for administrators, allowing them to create, update, and delete categories. Domain Models:

Added Category entity with id, name, and description fields. Modified Book class to include a set of categories. Other Changes:

Implemented CategoryRepository interface using JpaRepository. Implemented DTO classes for the Category entity.
Modified BookMapper class to include methods for mapping categories. Added CategoryMapper class with methods for mapping Category entities. Implemented CategoryService interface and CategoryServiceImpl class with methods for category management. Added CategoryController class with methods for handling category-related operations. General Requirements:

Used Liquibase for database schema management.
Implemented soft delete approach.
Added pagination, sorting, and Swagger documentation to all controllers. Endpoints:

Added endpoints for category management, including creating, retrieving, updating, and deleting categories, as well as retrieving books by category. Security Requirements:

Specified security requirements for non-authenticated users, users with role USER, and users with role ADMIN.

### CODE REVIEW

File: src/main/java/com/bookappstore/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    @GetMapping
+    @PreAuthorize("hasAnyRole('ADMIN', 'USER')")
+    public List<CategoryDto> getAll(Pageable pageable) {
+        return categoryService.findAll(pageable);
+    }
+
+    @Operation(summary = "Get category by ID",
+            description = "Get category by its ID")
+    @GetMapping("/{id}")
+    public CategoryDto getCategoryById(@PathVariable Long id) {
```
Comment (MAMentorFX) [inline src/main/java/com/bookappstore/controller/CategoryController.java:50]: @PreAuthorize?

File: src/main/java/com/bookappstore/dto/book/BookDtoWithoutCategoryIds.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+@Data
+public class BookDtoWithoutCategoryIds {
+    private Long id;
+    private String title;
+    private String author;
+    private String isbn;
+    private BigDecimal price;
+    private String description;
+    private String coverImage;
+
```
Comment (MAMentorFX) [inline src/main/java/com/bookappstore/dto/book/BookDtoWithoutCategoryIds.java:15]:
```suggestion
```

File: src/main/java/com/bookappstore/repository/book/spec/BookSpecificationBuilder.java [hunk 3]
```diff
@@ -69,6 +69,12 @@ public Specification<Book> build(BookSearchParametersDto bookSearchParametersDto
                     .getSpecification(bookSearchParametersDto.description()));
         }
 
+        if (bookSearchParametersDto.categories() != null && bookSearchParametersDto.categories()
+                .length > 0) {
+            specification = specification.and(bookSpecificationProviderManager
+                    .getSpecificationProvider("category")
+                    .getSpecification(bookSearchParametersDto.categories()));
+        }
```
Comment (MAMentorFX) [inline src/main/java/com/bookappstore/repository/book/spec/BookSpecificationBuilder.java:63]: I dont like how your  build method works. There's too many duplicates. Think of some abstraction

File: src/main/java/com/bookappstore/service/impl/BookServiceImpl.java [hunk 4]
```diff
@@ -21,16 +25,20 @@ public class BookServiceImpl implements BookService {
...
+        book.setCategories(categories);
         return bookMapper.toDto(bookRepository.save(book));
     }
 
     @Override
     public List<BookDto> findAll(Pageable pageable) {
-        return bookRepository.findAll(pageable).stream()
+        return bookRepository.findAllWithCategories(pageable).stream()
                 .map(bookMapper::toDto)
                 .toList();
```
Comment (MAMentorFX) [inline src/main/java/com/bookappstore/service/impl/BookServiceImpl.java]: Mapstruct supports mapping collections from the box

File: src/main/java/com/bookappstore/service/impl/BookServiceImpl.java [hunk 5]
```diff
@@ -21,16 +25,20 @@ public class BookServiceImpl implements BookService {
     private final BookRepository bookRepository;
     private final BookMapper bookMapper;
     private final BookSpecificationBuilder bookSpecificationBuilder;
+    private final CategoryRepository categoryRepository;
 
     @Override
     public BookDto save(CreateBookRequestDto createBookRequestDto) {
```
Comment (MAMentorFX) [inline src/main/java/com/bookappstore/service/impl/BookServiceImpl.java:33]: @Transactional. You make more than one call to DB

File: src/main/java/com/bookappstore/service/impl/BookServiceImpl.java [hunk 6]
```diff
@@ -59,12 +67,22 @@ public List<BookDto> search(BookSearchParametersDto params) {
 
     @Override
     public BookDto update(Long id, CreateBookRequestDto requestDto) {
```
Comment (MAMentorFX) [inline src/main/java/com/bookappstore/service/impl/BookServiceImpl.java:70]: @Transactional. You make more than one call to DB

File: src/main/java/com/bookappstore/service/impl/BookServiceImpl.java [hunk 7]
```diff
@@ -59,12 +67,22 @@ public List<BookDto> search(BookSearchParametersDto params) {
...
         Book updatedBook = bookRepository.save(bookById);
         return bookMapper.toDto(updatedBook);
     }
+
+    @Override
+    public List<BookDtoWithoutCategoryIds> findAllByCategoryId(Long categoryId) {
+        return bookRepository.findAllByCategoryId(categoryId)
+                .stream()
+                .map(bookMapper::toDtoWithoutCategories)
+                .toList();
```
Comment (MAMentorFX) [inline src/main/java/com/bookappstore/service/impl/BookServiceImpl.java]: Mapstruct supports mapping collections from the box

File: src/main/java/com/bookappstore/service/impl/CategoryServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    public CategoryDto save(CategoryDto requestDto) {
+        Category category = categoryMapper.toEntity(requestDto);
+        return categoryMapper.toDto(categoryRepository.save(category));
+    }
+
+    @Override
+    public List<CategoryDto> findAll(Pageable pageable) {
+        return categoryRepository.findAll(pageable).stream()
+                .map(categoryMapper::toDto)
+                .toList();
```
Comment (MAMentorFX) [inline src/main/java/com/bookappstore/service/impl/CategoryServiceImpl.java:30]: Mapstruct supports mapping collections from the box

File: src/main/java/com/bookappstore/service/impl/CategoryServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    @Override
+    public CategoryDto getById(Long id) {
+        Category category = categoryRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Can't find category by id: " + id)
+        );
+        return categoryMapper.toDto(category);
+    }
+
+    @Override
+    public CategoryDto update(Long id, CategoryDto categoryDto) {
```
Comment (MAMentorFX) [inline src/main/java/com/bookappstore/service/impl/CategoryServiceImpl.java:42]: @Transactional. You make more than one call to DB

File: src/main/java/com/bookappstore/service/impl/CategoryServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        );
+        categoryById.setName(categoryDto.getName());
+        categoryById.setDescription(categoryDto.getDescription());
+        return categoryMapper.toDto(categoryRepository.save(categoryById));
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        categoryRepository.deleteById(id);
+    }
```
Comment (MAMentorFX) [inline src/main/java/com/bookappstore/service/impl/CategoryServiceImpl.java:54]: @Transactional required. It wont save the entity without it

File: src/main/resources/db/changelog/changes/10-create-categories-table.yaml [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+                    unique: true
+              - column:
+                  name: description
+                  type: varchar(255)
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  constraints:
+                    nullable: false
+                  defaultValue: "false"
```
Comment (MAMentorFX) [inline src/main/resources/db/changelog/changes/10-create-categories-table.yaml:30]:
```suggestion
                  defaultValue: "false"
                  
```

File: src/main/resources/db/changelog/changes/11-fill-categories-table.yaml [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+        - insert:
+            tableName: categories
+            columns:
+              - column: {name: "name", value: "Fiction"}
+              - column: {name: "description", value: "Fiction books"}
+        - insert:
+            tableName: categories
+            columns:
+              - column: {name: "name", value: "Classic"}
+              - column: {name: "description", value: "Classic books"}
```
Comment (MAMentorFX) [inline src/main/resources/db/changelog/changes/11-fill-categories-table.yaml:15]:
```suggestion
              - column: {name: "description", value: "Classic books"}
              
```

File: src/main/resources/db/changelog/changes/12-create-books-categories-table.yaml [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+                    primaryKey: true
+              - column:
+                  name: book_id
+                  type: bigint
+                  constraints:
+                    foreignKeyName: fk_books_categories_book
+                    referencedTableName: books
+                    referencedColumnNames:  id
+                    nullable:  false
+                    primaryKey: true
```
Comment (MAMentorFX) [inline src/main/resources/db/changelog/changes/12-create-books-categories-table.yaml:26]:
```suggestion
                    primaryKey: true
                    
```

File: src/main/resources/db/changelog/changes/13-fill-books-categories-table.yaml [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+        - insert:
+            tableName: books_categories
+            columns:
+              - column: {name: "category_id", value: "1"}
+              - column: {name: "book_id", value: "2"}
+        - insert:
+            tableName: books_categories
+            columns:
+              - column: {name: "category_id", value: "2"}
+              - column: {name: "book_id", value: "3"}
```
Comment (MAMentorFX) [inline src/main/resources/db/changelog/changes/13-fill-books-categories-table.yaml:25]:
```suggestion
              - column: {name: "book_id", value: "3"}
              
```

**Review by MAMentorFX — CHANGES_REQUESTED** [review]

File: src/main/java/com/bookappstore/repository/book/spec/BookSpecificationBuilder.java (hunk 3 above)
Comment (VKhmil) [reply src/main/java/com/bookappstore/repository/book/spec/BookSpecificationBuilder.java:63]: Thanks, good tip

File: src/main/java/com/bookappstore/controller/CategoryController.java (hunk 1 above)
Comment (VKhmil) [reply src/main/java/com/bookappstore/controller/CategoryController.java:50]: Forgot about that

**Review by Vasul-Pankiv — APPROVED** [review]
good Job!

**Pull request merged**